src/transformers/models/rt_detr/modeling_rt_detr.py

```diff
@@ -0,0 +1,1652 @@
+            wh = torch.ones_like(grid_xy) * grid_size * (2.0**lvl)
+            anchors.append(torch.concat([grid_xy, wh], -1).reshape(-1, h * w, 4))
+
+        anchors = torch.concat(anchors, 1).to(device)
+        valid_mask = ((anchors > EPS) * (anchors < 1 - EPS)).all(-1, keepdim=True)
+        anchors = torch.log(anchors / (1 - anchors))
+        anchors = torch.where(valid_mask, anchors, torch.inf)
+
+        return anchors, valid_mask
+
+    def forward(self, feats, targets=None):
+        # get projection features
+        projected_features = [self.input_proj[i](feat) for i, feat in enumerate(feats)]
+        if self.num_levels > len(projected_features):
+            len_srcs = len(projected_features)
+            for i in range(len_srcs, self.num_levels):
+                if i == len_srcs:
+                    projected_features.append(self.input_proj[i](feats[-1]))
+                else:
+                    projected_features.append(self.input_proj[i](projected_features[-1]))
+
+        # get encoder inputs
+        feat_flatten = []
+        spatial_shapes = []
+        level_start_index = [0]
+        for feat in projected_features:
+            height, width = feat.shape[-2:]
+            # [batch, channels, height, width] -> [batch, height*width, channel]
+            feat_flatten.append(feat.flatten(2).permute(0, 2, 1))
+            # [num_levels, 2]
+            spatial_shapes.append([height, width])
+            # [level], start index of each level
+            level_start_index.append(height * width + level_start_index[-1])
+
+        # [batch, level, channel]
+        feat_flatten = torch.concat(feat_flatten, 1)
+        level_start_index.pop()
+
+        # prepare denoising training
+        if self.training and self.num_denoising > 0 and targets is not None:
+            denoising_class, denoising_bbox_unact, attn_mask, dn_meta = get_contrastive_denoising_training_group(
+                targets,
+                self.num_classes,
+                self.num_queries,
+                self.denoising_class_embed,
+                num_denoising=self.num_denoising,
+                label_noise_ratio=self.label_noise_ratio,
+                box_noise_scale=self.box_noise_scale,
+            )
+        else:
+            denoising_class, denoising_bbox_unact, attn_mask, dn_meta = None, None, None, None
+
+        batch_size = len(feat_flatten)
+        device = feat_flatten.device
+
+        # prepare input for decoder
+        if self.training or self.eval_spatial_size is None:
+            anchors, valid_mask = self.generate_anchors(spatial_shapes, device=device)
+        else:
+            anchors, valid_mask = self.anchors.to(device), self.valid_mask.to(device)
+
+        memory = valid_mask.to(feat_flatten.dtype) * feat_flatten
+
+        output_memory = self.enc_output(memory)
+
+        enc_outputs_class = self.enc_score_head(output_memory)
+        enc_outputs_coord_unact = self.enc_bbox_head(output_memory) + anchors
```

Comment:
What is "unact"?

Reply:
It seems that it stands for "unactivated", meaning it wasn't passed by the activation function (sigmoid) yet